### DISCUSSION TITLE
How to set sparsity pattern and relaxation steps for ISAI preconditioner?

### DISCUSSION THREAD

**Discussion by learning-chip** [ginkgo-project/ginkgo#1119] [Q&A] [answered]
I saw the paper [Incomplete Sparse Approximate Inverses for Parallel Preconditioning](https://www.sciencedirect.com/science/article/abs/pii/S016781911730176X) along with its implementation at [`gko::preconditioner::Isai`](https://ginkgo-project.github.io/ginkgo-generated-documentation/doc/develop/classgko_1_1preconditioner_1_1Isai.html) (thanks for releasing for code!). I find the idea of ISAI very interesting and want to reproduce some results in the paper.

However, there doesn't seem a way to change the sparsity pattern (using L^2 pattern for example) and relaxation steps (Eq 15 of the paper) as did in the paper.

<img width="726" alt="截屏2022-09-13 下午9 36 25" src="https://user-images.githubusercontent.com/80731350/189915743-6f9f8d5f-e558-442e-b0f0-a765371a0dd4.png">

I think what's implemented in Ginkgo is referred as "SpMV SAI" in the paper, which does not use relaxation steps. 

I'd like to produce something like this Table 4 in the paper that sweeps over ISAI parameters:

<details>

<img width="478" alt="截屏2022-09-13 下午9 10 03" src="https://user-images.githubusercontent.com/80731350/189915992-c0707714-3601-4f2b-9421-d19183b0db8e.png">

</details>

**Comment by learning-chip** [comment]
In the unit test `reference/test/preconditioner/isai_kernels` I found the call `with_sparsity_power()` that seems to set the sparsity pattern:

https://github.com/ginkgo-project/ginkgo/blob/693bc4fa083a09c1f737bf69e7ff69844f2204cd/reference/test/preconditioner/isai_kernels.cpp#L1144-L1149

As for the relaxation scheme, I could extract the inverse factor and manually type the formula (Eq 15.). Or is there a more efficient implementation available?

**Comment by fritzgoebel** [comment, marked as answer]
Hi, thanks for reaching out!
You seem to have already found the sparsity power parameter, that's great. For the relaxation steps, you can have a look at the example `ir-ilu-preconditioned-solver`. In line 120, there is a triangular solver defined as iterative refinement with block-Jacobi as the inner solver. If you replace the block-Jacobi with your ISAI factory, you should end up with what you are looking for. That would look something like 
```
    auto isai_factory =
        gko::share(gko::preconditioner::LowerIsai<value_type, index_type>::build() 
                       .with_sparsity_power(2) 
                       .on(this->exec));

    auto trisolve_factory =
        gko::solver::Ir<value_type>::build()
            .with_solver(isai_factory)
            .with_criteria(
                gko::stop::Iteration::build().with_max_iters(20u).on(exec))
            .on(exec);
```
I hope this helps!

**Reply by learning-chip** [reply]
Thanks, composing `gko::solver::Ir` with ISAI is brilliant!
